Content/Items/Breacher/Projectiles.BreachCannonSentry.cs

```diff
@@ -0,0 +1,585 @@
+				sb.Draw(tex, superLaserEndpoint - Main.screenPosition, null, blueCyan, 0, tex.Size() / 2, 0.35f * superLaserSizeMult, SpriteEffects.None, 0f);
+			}
+
+			DrawBalls(sb, BlendState.Additive, Color.White, Color.White, 0.35f);
+		}
+
+		private void SpawnParticles()
+		{
+			Vector2 direction = laserEndpoint.DirectionTo(laserStartpoint);
+
+			Color color1 = Color.Cyan;
+			Color color2 = new Color(0, 0, 255);
+
+			for (int i = 0; i < 5; i++)
+				Dust.NewDustPerfect(laserEndpoint, ModContent.DustType<BreachImpactGlow>(), direction.RotatedByRandom(0.6f) * Main.rand.NextFloat(12), 0, Color.Lerp(color1, color2, Main.rand.NextFloat()), Main.rand.NextFloat(0.25f, 0.6f));
+
+			for (int i = 0; i < 3; i++)
+			{
+				Vector2 vel = direction.RotatedByRandom(0.6f) * Main.rand.NextFloat(9);
+				Dust.NewDustPerfect(laserEndpoint + new Vector2(0, 35) + (vel * 2), ModContent.DustType<BreachImpactSpark>(), vel, 0, Color.Lerp(color1, color2, Main.rand.NextFloat()), Main.rand.NextFloat(1.25f, 1.6f));
+			}
+
+			if (superLaser)
+            {
+				direction = superLaserEndpoint.DirectionTo(superLaserStartpoint);
+
+				for (int i = 0; i < 5; i++)
+					Dust.NewDustPerfect(superLaserEndpoint, ModContent.DustType<BreachImpactGlow>(), direction.RotatedByRandom(0.6f) * Main.rand.NextFloat(12), 0, Color.Lerp(color1, color2, Main.rand.NextFloat()), Main.rand.NextFloat(0.25f, 0.6f));
+
+				for (int i = 0; i < 3; i++)
+				{
+					Vector2 vel = direction.RotatedByRandom(0.6f) * Main.rand.NextFloat(9);
+					Dust.NewDustPerfect(superLaserEndpoint + new Vector2(0, 35) + (vel * 2), ModContent.DustType<BreachImpactSpark>(), vel, 0, Color.Lerp(color1, color2, Main.rand.NextFloat()), Main.rand.NextFloat(1.25f, 1.6f));
+				}
+			}
+		}
+	}
+	public class BreachImpactGlow : Dusts.Glow
+	{
+		public override void OnSpawn(Dust dust)
+		{
+			dust.noGravity = true;
+			dust.frame = new Rectangle(0, 0, 64, 64);
+
+			dust.shader = new Terraria.Graphics.Shaders.ArmorShaderData(new Ref<Effect>(StarlightRiver.Instance.Assets.Request<Effect>("Effects/GlowingDust", ReLogic.Content.AssetRequestMode.ImmediateLoad).Value), "GlowingDustPass");
+			int a = 1; //nessecary for data setting reasons I think?
+		}
+		public override bool Update(Dust dust)
+        {
+			dust.scale *= 0.85f;
+			return base.Update(dust);
+        }
+	}
+	class BreachImpactSpark : Dusts.BuzzSpark
+    {
+		public override void OnSpawn(Dust dust)
+		{
+			dust.fadeIn = 0;
+			dust.noLight = false;
+			dust.frame = new Rectangle(0, 0, 5, 50);
+
+			dust.shader = new Terraria.Graphics.Shaders.ArmorShaderData(new Ref<Effect>(StarlightRiver.Instance.Assets.Request<Effect>("Effects/ShrinkingDust").Value), "ShrinkingDustPass");
+		}
+		public override bool Update(Dust dust)
+		{
+			dust.fadeIn++;
+			return base.Update(dust);
+		}
+	}
+
+	class BreacherLaserUpdater : ModSystem
+    {
+        public override void PreUpdateProjectiles()
```

Comment:
This seems like a very hacky workaround -- what is the purpose of this class? Is there not a better way to reset the projectile from within it's own class? What about the preAI hook?

Reply:
The problem with PreAI is it has to reset every single breacher laser before even one AI is called.